daemon/api_system_volumes.go

```diff
@@ -21,23 +21,167 @@ package daemon
 
 import (
 	"encoding/json"
+	"errors"
+	"fmt"
 	"net/http"
+	"net/url"
 
+	"github.com/snapcore/snapd/asserts"
+	"github.com/snapcore/snapd/client"
+	"github.com/snapcore/snapd/gadget"
 	"github.com/snapcore/snapd/overlord/auth"
+	"github.com/snapcore/snapd/overlord/devicestate"
 	"github.com/snapcore/snapd/overlord/fdestate"
+	"github.com/snapcore/snapd/overlord/snapstate"
+	"github.com/snapcore/snapd/overlord/state"
 	"github.com/snapcore/snapd/secboot/keys"
+	"github.com/snapcore/snapd/strutil"
 )
 
 var systemVolumesCmd = &Command{
 	Path:        "/v2/system-volumes",
+	GET:         getSystemVolumes,
 	POST:        postSystemVolumesAction,
+	ReadAccess:  rootAccess{},
 	WriteAccess: rootAccess{},
 }
 
 var (
 	fdestateReplaceRecoveryKey = fdestate.ReplaceRecoveryKey
+	fdeMgrGenerateRecoveryKey  = (*fdestate.FDEManager).GenerateRecoveryKey
+	fdeMgrCheckRecoveryKey     = (*fdestate.FDEManager).CheckRecoveryKey
+
+	snapstateGadgetInfo = snapstate.GadgetInfo
 )
 
+func parseSystemVolumesOptionsFromURL(q url.Values) (opts *client.SystemVolumesOptions, err error) {
+	opts = &client.SystemVolumesOptions{
+		ContainerRoles: q["container-role"],
+	}
+	switch q.Get("by-container-role") {
+	case "true", "false", "":
+		opts.ByContainerRole = q.Get("by-container-role") == "true"
+	default:
+		return nil, errors.New(`"by-container-role" query parameter when used must be set to "true" or "false" or left unset`)
+	}
+	if len(opts.ContainerRoles) > 0 && opts.ByContainerRole {
+		return nil, errors.New(`"container-role" query parameter conflicts with "by-container-role"`)
+	}
+	return opts, nil
+}
+
+func getCurrentGadgetAndModel(st *state.State) (*gadget.Info, *asserts.Model, error) {
+	deviceCtx, err := snapstate.DeviceCtx(st, nil, nil)
+	if err != nil {
+		return nil, nil, fmt.Errorf("cannot get device context: %v", err)
+	}
+	gadgetInfo, err := snapstateGadgetInfo(st, deviceCtx)
+	if err != nil {
+		return nil, nil, fmt.Errorf("cannot get gadget snap info: %v", err)
+	}
+	gadgetDir := gadgetInfo.MountDir()
+
+	model := deviceCtx.Model()
+
+	info, err := gadget.ReadInfoAndValidate(gadgetDir, model, nil)
+	if err != nil {
+		return nil, nil, fmt.Errorf("cannot read gadget: %v", err)
+	}
+	return info, model, nil
+}
+
+func getGadgetVolumeStructuresWithKeyslots(st *state.State, fdemgr *fdestate.FDEManager) ([]devicestate.VolumeStructureWithKeyslots, error) {
+	st.Lock()
+	defer st.Unlock()
+
+	gadgetInfo, _, err := getCurrentGadgetAndModel(st)
+	if err != nil {
+		return nil, fmt.Errorf("cannot get current gadget: %v", err)
+	}
+
+	structures, err := devicestate.GetGadgetVolumeStructuresWithKeyslots(fdemgr, gadgetInfo)
```

Comment:
it's strange for this to take the gadgetnfo, deviceState should be able to find by itself, no?

Reply:
updated